core/shared/src/main/scala-3/org/virtuslab/yaml/YamlDecoderCrossCompat.scala

```diff
@@ -94,17 +36,153 @@ private[yaml] trait DecoderMacros {
+          }
+          val namedInstance = '{ (${ fieldName }, $fieldFormat, ${ isOption }) }
+          namedInstance :: prepareInstances(Type.of[labelsTail], Type.of[tpesTail])
+
+    p match
+      case '{
+            $m: Mirror.ProductOf[T] {
+              type MirroredElemLabels = elementLabels; type MirroredElemTypes = elementTypes
+            }
+          } =>
+        val allInstancesExpr =
+          Expr.ofList(prepareInstances(Type.of[elementLabels], Type.of[elementTypes]))
+        val defaultParamsExpr = findDefaultParams[T]
+
+        '{
+          new YamlDecoder[T] {
+            private val allInstances  = $allInstancesExpr
+            private val defaultParams = $defaultParamsExpr
+            private val mirror        = $p
+
+            override def construct(node: Node)(using
+                constructor: LoadSettings = LoadSettings.empty
+            ): Either[ConstructError, T] =
+              node match
+                case Node.MappingNode(mappings, _) =>
+                  for {
+                    valuesMap <- extractKeyValues(mappings)
+                    constructedValues <- constructValues(
+                      allInstances,
+                      valuesMap,
+                      defaultParams,
+                      mirror,
+                      node
+                    )
+                  } yield (constructedValues)
+                case _ =>
+                  Left(
+                    ConstructError.from(
+                      s"Expected MappingNode, got ${node.getClass.getSimpleName}",
+                      node
+                    )
+                  )
+          }
+        }
+
+  protected def findDefaultParams[T](using
+      quotes: Quotes,
+      tpe: Type[T]
+  ): Expr[Map[String, () => Any]] =
+    import quotes.reflect.*
+
+    TypeRepr.of[T].classSymbol match
+      case None => '{ Map.empty[String, () => Any] }
+      case Some(sym) =>
+        val comp = sym.companionClass
+        try
+          val mod = Ref(sym.companionModule)
+          val names =
+            for p <- sym.caseFields if p.flags.is(Flags.HasDefault)
+            yield p.name
+          val namesExpr: Expr[List[String]] =
+            Expr.ofList(names.map(Expr(_)))
+
+          val body = comp.tree.asInstanceOf[ClassDef].body
+          val idents: List[Ref] =
+            for
+              case deff @ DefDef(name, _, _, _) <- body
+              if name.startsWith("$lessinit$greater$default")
+            yield mod.select(deff.symbol)
+          val typeArgs = TypeRepr.of[T].typeArgs
+          val identsExpr: Expr[List[() => Any]] =
+            if typeArgs.isEmpty then
+              Expr.ofList(idents.map(_.asExpr).map { case '{ $x } => '{ () => $x } })
+            else
+              Expr.ofList(
+                idents.map(_.appliedToTypes(typeArgs).asExpr).map { case '{ $x } => '{ () => $x } }
+              )
 
+          '{ $namesExpr.zip($identsExpr).toMap }
```

Comment:
I'm not sure if the order of declarations in `body` is reliable. Maybe it would be better to use something like `sym.companionClass.methodMember("$lessinit$greater$default$" + n.toString).head` to get the default value provider method for the n-th field?